src/nifs/protogalaxy/pow_i.rs

```diff
@@ -0,0 +1,123 @@
+use std::{iter, num::NonZeroUsize};
+
+use bitter::{BitReader, LittleEndianReader};
+use ff::PrimeField;
+use tracing::*;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Input `t` is not enought for represent `i`")]
+    TooLittleT {
+        t: NonZeroUsize,
+        representation_lenght: NonZeroUsize,
+    },
+}
+
+fn pow_i<F: PrimeField>(
+    i: usize,
+    t: NonZeroUsize,
+    challenges_powers: impl Iterator<Item = F>,
+) -> F {
+    let bytes = i.to_le_bytes();
+    let mut reader = LittleEndianReader::new(&bytes);
+
+    iter::repeat_with(|| reader.read_bit().unwrap_or(false))
+        .zip(challenges_powers)
+        .map(|(b_j, beta_in_2j)| match b_j {
+            true => beta_in_2j,
+            false => F::ONE,
+        })
+        .take(t.get())
+        .reduce(|acc, coeff| acc * coeff)
+        .unwrap()
+}
+
+/// Constructs an iterator yielding evaluated `pow_i` for indices from 0 to `n`.
+///
+/// # Params
+/// - `t` - lenght of binary representation of `i`. If `t` is too small to represent `i`, then
+///       [`PowError::TooLittleT`] will return.
+/// - `n` - limit for `pow_i` relations
+/// - `challenge` - challenge for `pow_i` relation
+///
+/// # Mathematical Representation
+///
+/// For an integer `i` ranging from 0 to `n`, its binary representation `[i]_2 = (b_0, b_1, ..., b_{t-1})`,
+/// and an input expression `challenge`, the polynomial term `pow_i` is defined as:
+///
+/// `beta` is `challenge`
+/// ```math
+/// pow_i(\boldsymbol{\beta}) =
+///     pow_i(\beta, \cdots, \beta^{2^{t-1}}) =
+///     \prod_{j=0}^{t-1}(1 - b_j + b_j \cdot \beta^{2^j})
+/// ```
+/// where `b_j` is the j-th bit of the binary representation of the integer `i`.
+pub fn iter_eval_of_pow_i<F: PrimeField>(
+    t: NonZeroUsize,
+    n: NonZeroUsize,
+    challenge: F,
+) -> Result<impl Iterator<Item = F>, Error> {
+    let n = n.get();
+
+    if t.get() < n.leading_ones() as usize {
+        return Err(Error::TooLittleT {
+            t,
+            representation_lenght: NonZeroUsize::new(n.leading_ones() as usize).unwrap(),
+        });
+    }
+
+    let challenges = iter::successors(Some(challenge), |v| Some(v.square()))
+        .take(t.get())
+        .collect::<Box<[_]>>();
+
+    Ok((0..=n).map(move |i| pow_i(i, t, challenges.iter().copied())))
```

Comment:
why use copied instead of move?

Reply:
didn't get this part: "there was no sense to copy inside and add lifetime, now there is". but since you have removed the copied. will just approve the PR

Reply:
Before it was `1 - b + b * (*beta)` inside and so there were copies happening at each iteration (deref makes a copy inside) and it didn't make sense to do it inside the fn, it made sense to copy the iterator at once.

But with `match` the situation changes, only some elements are copied. 

`move` doesn't play a role here, we pass the collection ownership inside the closure in any case, the question is how the values from the closure get into the fn